src/imageproxy.rs

```diff
@@ -410,21 +429,42 @@ impl ImageProxy {
                     error: reply.error.into(),
                 });
             }
-            let fdret = match (fdret, reply.pipeid) {
-                (Some(fd), n) => {
-                    if n == 0 {
-                        return Err(Error::Other("got fd but no pipeid".into()));
-                    }
-                    Some((fd, n))
+            let first_fd = fdret.next();
+            let second_fd = fdret.next();
+            if fdret.next().is_some() {
+                return Err(Error::Other(
+                    format!("got more than two file descriptors").into(),
+                ));
+            }
+            let pipeid = NonZeroU32::new(reply.pipeid);
+            let fdret = match (first_fd, second_fd, pipeid) {
+                // No fds, no pipeid
```

Comment:
I have a pattern I like to use in these cases, something like:

```rust
let fds = mything.filter_map(...).collect::<Vec<_>>();
let pipeid = reply.pipeid;

match (pipeid, fds.as_slice()) {
  (&[data_fd], Some(pipeid)) => { good case },
  (&[data_fd, err_fd], None) => { good case },
  ... => { base cases }
}
```

it's slightly gross because it allocates the Vec, but I don't think that matters a whole lot in this case, and the result is a lot easier to read... It also means that you don't need to handle the impossible case of the iterator having returned None the first time and Some the second.

Reply:
I ended up splitting up the "parsing" here into a separate helper function and added unit tests, which I think fixes this readability.